File: python/techai_webutils/tests/unit/clients_job_event_stack_test.py
"""Tests for the Python Job + EventHandler decorator stacks.

Parity with Go's ``go/tests/unit/clients_jobevent_stack_test.go``: the event stack skips a
duplicate and dead-letters a message that still fails after retries; the job stack
runs only on the leader.
"""

from __future__ import annotations

import pytest
from techai_webutils.clients.decorators.event_stack import EventStackDeps, wrap_handler
from techai_webutils.clients.decorators.job_stack import JobStackDeps, wrap_job
from techai_webutils.core.errors.errors import UnavailableError
from techai_webutils.core.interfaces.messaging import Message
from techai_webutils.foundation.resilience.dedup import MemoryDeduplicator
from techai_webutils.foundation.resilience.dlq import DeadLetterQueue, StubDeadLetterBackend
from techai_webutils.foundation.resilience.leader import AlwaysLeader


class _NotLeader:
    """A ``LeaderElector`` that never holds leadership (for the leader-gating test)."""

    async def is_leader(self) -> bool:
        """Never the leader."""
        return False


def _message(message_id: str = "m1") -> Message:
    """Build a minimal message (id, topic, payload) for the stack tests."""
    return Message(id=message_id, topic="t", payload=b"x")


@pytest.mark.asyncio
async def test_wrap_handler_dedup_skips_duplicate() -> None:
    """The event stack runs a redelivered message's handler exactly once.

    Why this test is important:
        - Under at-least-once delivery the same message can arrive twice; the dedup layer is the
          guard, and re-running the handler would double-process (e.g. duplicate side effects).

    What it tests:
        - Two handler invocations for the same message id run the inner handler exactly once.
    """
    calls = 0

    async def inner(_msg: Message) -> None:
        nonlocal calls
        calls += 1

    handler = wrap_handler(inner, EventStackDeps(dedup=MemoryDeduplicator()))
    msg = _message()

    await handler(msg)
    await handler(msg)

    assert calls == 1, "the handler must run exactly once for a duplicate delivery"


@pytest.mark.asyncio
async def test_wrap_handler_retries_then_dead_letters() -> None:
    """A message that still fails after retries is dead-lettered and acked.

    Why this test is important:
        - A poison message that neither dead-letters nor acks would loop forever; retrying the
          transient failure then routing it to the DLQ is how the consumer makes progress without
          losing the message.

    What it tests:
        - An always-failing (transient) handler is retried up to the configured attempts and then
          recorded in the dead-letter backend, with no exception escaping (the message is acked).
    """
    backend = StubDeadLetterBackend()
    attempts = 0

    async def inner(_msg: Message) -> None:
        nonlocal attempts
        attempts += 1
        raise UnavailableError("downstream down")

    handler = wrap_handler(
        inner,
        EventStackDeps(dlq=DeadLetterQueue(backend), retry_max_attempts=2),
    )

    await handler(_message())  # must not raise -- the failure is dead-lettered

    assert attempts == 2, "the transient failure is retried before dead-lettering"
    assert len(backend.letters) == 1
    assert backend.letters[0].id == "m1"


@pytest.mark.asyncio
async def test_wrap_job_leader_gating() -> None:
    """The job stack runs the job only on the leader, so an N-replica schedule fires once.

    Why this test is important:
        - Running a scheduled job on every replica would duplicate its effect (e.g. N reaper
          sweeps); leader election is the guard that makes it fire once.

    What it tests:
        - A non-leader instance skips the job (never runs it); the leader runs it exactly once.
    """
    follower_calls = 0
    leader_calls = 0

    async def follower_job() -> None:
        nonlocal follower_calls
        follower_calls += 1

    async def leader_job() -> None:
        nonlocal leader_calls
        leader_calls += 1

    await wrap_job(follower_job, JobStackDeps(leader=_NotLeader()))()
    assert follower_calls == 0, "a non-leader must not run the job"

    await wrap_job(leader_job, JobStackDeps(leader=AlwaysLeader()))()
    assert leader_calls == 1, "the leader runs the job once"


@pytest.mark.asyncio
async def test_wrap_handler_open_breaker_fails_fast_without_starting_the_handler() -> None:
    """With the circuit breaker open, the handler is not even called (no orphaned coroutine).

    Why this test is important:
        - The attempt used to build the handler coroutine before entering the breaker; when the breaker
          was open it raised first, so every message left an un-awaited coroutine behind (a
          "coroutine was never awaited" RuntimeWarning and a leaked frame per message while open).

    What it tests:
        - After one failure opens a threshold-1 breaker, the next message raises CircuitOpenError, the
          handler callable is not invoked again, and no "never awaited" RuntimeWarning is emitted.
    """
    import gc
    import warnings

    from techai_webutils.foundation.resilience.circuit_breaker import CircuitBreaker, CircuitOpenError

    calls = 0

    async def failing() -> None:
        raise UnavailableError("downstream down")

    def inner(_msg: Message) -> object:
        nonlocal calls
        calls += 1
        return failing()

    handler = wrap_handler(
        inner,  # type: ignore[arg-type]
        EventStackDeps(circuit_breaker=CircuitBreaker(failure_threshold=1, recovery_timeout=60.0)),
    )
    with pytest.raises(UnavailableError):
        await handler(_message("m1"))  # opens the breaker

    with warnings.catch_warnings(record=True) as caught:
        warnings.simplefilter("always")
        with pytest.raises(CircuitOpenError):
            await handler(_message("m2"))
        gc.collect()

    assert calls == 1
    assert not [w for w in caught if "never awaited" in str(w.message)]

File: go/tests/unit/clients_jobevent_stack_test.go
package unit_test

import (
	"context"
	"errors"
	"sync/atomic"
	"testing"

	"github.com/stretchr/testify/require"

	jobdec "github.com/gt-tech-ai/knowledge-engine/go/clients/jobs/decorators"
	msgdec "github.com/gt-tech-ai/knowledge-engine/go/clients/messaging/decorators"
	"github.com/gt-tech-ai/knowledge-engine/go/core/interfaces"
	"github.com/gt-tech-ai/knowledge-engine/go/foundation/resilience/deadletter"
	"github.com/gt-tech-ai/knowledge-engine/go/foundation/resilience/dedup"
	"github.com/gt-tech-ai/knowledge-engine/go/foundation/resilience/leader"
	"github.com/gt-tech-ai/knowledge-engine/go/tests/fixtures"
	"github.com/gt-tech-ai/knowledge-engine/go/tests/mocks"
	"go.uber.org/mock/gomock"
)

// TestWrapHandler_DedupSkipsDuplicate tests that the EventHandler stack skips a message
// whose key was already processed, so a redelivery is handled exactly once (D-40).
//
// Why this test is important:
//   - Under at-least-once delivery the same message can arrive twice; the dedup layer is
//     the guard, and re-running the handler would double-process (e.g. duplicate side effects).
//
// What it tests:
//   - Two handler invocations for the same message id run the inner handler exactly once.
func TestWrapHandler_DedupSkipsDuplicate(t *testing.T) {
	t.Parallel()

	var calls int32
	handler := msgdec.WrapHandler(
		func(context.Context, *interfaces.Message) error {
			atomic.AddInt32(&calls, 1)
			return nil
		},
		msgdec.EventStackDeps{Dedup: dedup.NewMemory(0), Name: "test"},
	)

	msg := &interfaces.Message{ID: "m1", Topic: "t"}
	require.NoError(t, handler(context.Background(), msg), "first delivery processed")
	require.NoError(t, handler(context.Background(), msg), "duplicate delivery skipped")
	require.Equal(t, int32(1), atomic.LoadInt32(&calls), "handler must run exactly once")
}

// TestWrapHandler_TerminalFailureDeadLettered tests that a message that still fails after
// the resilience stack is routed to the dead-letter queue and acked (D-34).
//
// Why this test is important:
//   - A poison message that neither dead-letters nor acks would loop forever; routing it to
//     the DLQ and acking is how the consumer makes progress without losing the message.
//
// What it tests:
//   - An always-failing handler (no retrier → one attempt) yields a nil return (acked) and the
//     dead-letter backend records the message.
func TestWrapHandler_TerminalFailureDeadLettered(t *testing.T) {
	t.Parallel()

	stub := &deadletter.StubBackend{}
	handler := msgdec.WrapHandler(
		func(context.Context, *interfaces.Message) error { return errors.New("boom") },
		msgdec.EventStackDeps{
			DeadLetter: deadletter.New(stub, fixtures.NewSpyLogger()),
			Name:       "test",
		},
	)

	msg := &interfaces.Message{ID: "m1", Topic: "t", Payload: []byte("x")}
	require.NoError(
		t,
		handler(context.Background(), msg),
		"a dead-lettered message is acked",
	)
	require.Len(t, stub.Letters, 1)
	require.Equal(t, "m1", stub.Letters[0].ID)
}

// TestWrapJob_LeaderGating tests that the Job stack runs the job only on the leader, so a
// job scheduled across N replicas fires once (D-30).
//
// Why this test is important:
//   - Running a scheduled job on every replica would duplicate its effect (e.g. N reaper
//     sweeps); leader election is the guard that makes it fire once.
//
// What it tests:
//   - A non-leader instance skips the job (never runs it); the leader runs it exactly once.
func TestWrapJob_LeaderGating(t *testing.T) {
	t.Parallel()

	var followerCalls, leaderCalls int32

	// notLeader is a generated LeaderElector mock that never holds leadership, so
	// the follower path must skip the job.
	notLeader := mocks.NewMockLeaderElector(gomock.NewController(t))
	notLeader.EXPECT().IsLeader(gomock.Any()).Return(false, nil).AnyTimes()

	follower := jobdec.WrapJob(
		func(context.Context) error { atomic.AddInt32(&followerCalls, 1); return nil },
		jobdec.JobStackDeps{Leader: notLeader, Name: "test"},
	)
	require.NoError(t, follower(context.Background()))
	require.Zero(t, atomic.LoadInt32(&followerCalls), "a non-leader must not run the job")

	elected := jobdec.WrapJob(
		func(context.Context) error { atomic.AddInt32(&leaderCalls, 1); return nil },
		jobdec.JobStackDeps{Leader: leader.AlwaysLeader{}, Name: "test"},
	)
	require.NoError(t, elected(context.Background()))
	require.Equal(
		t,
		int32(1),
		atomic.LoadInt32(&leaderCalls),
		"the leader runs the job once",
	)
}
